File: src/main/java/com/EmosewaPixel/pixellib/capabilities/ImprovedItemStackHandler.java
package com.EmosewaPixel.pixellib.capabilities;

import net.minecraft.item.ItemStack;
import net.minecraftforge.items.ItemStackHandler;

import java.util.Arrays;
import java.util.List;

public class ImprovedItemStackHandler extends ItemStackHandler {
    protected List<Integer> noInputSlots;
    protected List<Integer> noOutputSlots;

    public ImprovedItemStackHandler(int slots, Integer[] noInput, Integer[] noOutput) {
        super(slots);
        this.noInputSlots = Arrays.asList(noInput);
        this.noOutputSlots = Arrays.asList(noOutput);
    }

    public ImprovedItemStackHandler(int slots) {
        this(slots, null, null);
    }

    @Override
    public ItemStack insertItem(int slot, ItemStack stack, boolean simulate) {
        if (!isItemValid(slot, stack)) {
            return stack;
        }
        return super.insertItem(slot, stack, simulate);
    }

    @Override
    public boolean isItemValid(int slot, ItemStack itemStack) {
        return !noInputSlots.contains(slot);
    }

    @Override
    public ItemStack extractItem(int slot, int amount, boolean simulate) {
        if (noOutputSlots.contains(slot))
            return ItemStack.EMPTY;
        return super.extractItem(slot, amount, simulate);
    }
}

File: src/main/java/com/emosewapixel/pixellib/capabilities/ImprovedItemStackHandler.kt
package com.emosewapixel.pixellib.capabilities

import net.minecraft.item.ItemStack
import net.minecraftforge.items.ItemStackHandler

//Improved Item Stack Handlers are Item Stack Handlers that can blacklist slots for insertion and extraction
open class ImprovedItemStackHandler @JvmOverloads constructor(slots: Int, noInput: Array<Int>? = null, noOutput: Array<Int>? = null) : ItemStackHandler(slots) {
    protected var noInputSlots: List<Int> = listOf(*noInput!!)
    protected var noOutputSlots: List<Int> = listOf(*noOutput!!)

    override fun insertItem(slot: Int, stack: ItemStack, simulate: Boolean) = if (!isItemValid(slot, stack)) stack else super.insertItem(slot, stack, simulate)

    override fun isItemValid(slot: Int, itemStack: ItemStack) = !noInputSlots.contains(slot)

    override fun extractItem(slot: Int, amount: Int, simulate: Boolean) = if (noOutputSlots.contains(slot)) ItemStack.EMPTY else super.extractItem(slot, amount, simulate)
}